pkg/operator/clustermemberremovalcontroller/clustermemberremovalcontroller.go

```diff
@@ -0,0 +1,253 @@
+		isMachineAPIFunctionalFn:          isMachineAPIFunctional,
+		masterMachineSelector:             masterMachineSelector,
+		masterNodeSelector:                masterNodeSelector,
+		masterNodeLister:                  masterNodeLister,
+		masterMachineLister:               masterMachineLister,
+		networkLister:                     networkLister,
+		configMapListerForTargetNamespace: kubeInformersForNamespaces.InformersFor(operatorclient.TargetNamespace).Core().V1().ConfigMaps().Lister().ConfigMaps(operatorclient.TargetNamespace),
+		hasNetworkInformerSyncedFn:        hasNetworkInformerSyncedFn,
+	}
+	return factory.New().
+		WithSync(c.sync).
+		ResyncEvery(33*time.Minute). // make it slow since nodes are updated every few minutes
+		WithInformers(
+			masterNodeInformer,
+			masterMachineInformer,
+			kubeInformersForNamespaces.InformersFor(operatorclient.TargetNamespace).Core().V1().ConfigMaps().Informer(),
+		).ToController("ClusterMemberRemovalController", eventRecorder.WithComponentSuffix("cluster-member-removal-controller"))
+}
+
+func (c *ClusterMemberRemovalController) sync(ctx context.Context, _ factory.SyncContext) error {
+	if !c.hasNetworkInformerSyncedFn() {
+		return nil
+	}
+	// stop if the machine API is not functional
+	if isFunctional, err := c.isMachineAPIFunctionalFn(); err != nil {
+		return err
+	} else if !isFunctional {
+		return nil
+	}
+	etcdEndpointsConfigMap, err := c.configMapListerForTargetNamespace.Get("etcd-endpoints")
+	if err != nil {
+		return err // should not happen
+	}
+	if len(etcdEndpointsConfigMap.Data) == 0 {
+		return nil
+	}
+	for _, potentialMemberToRemoveIP := range etcdEndpointsConfigMap.Data {
+		memberLocator := potentialMemberToRemoveIP
+		machine, err := c.getMachineForMember(potentialMemberToRemoveIP)
+		if err != nil {
+			klog.Errorf("unable to get a machine for member: %v, err: %v", memberLocator, err)
+			continue
+		}
+		if machine != nil {
+			klog.V(4).Infof("cannot remove member: %v because a machine resource still exists (%v/%v)", memberLocator, machine.Name, machine.UID)
+			continue
+		}
+
+		node, err := c.getNodeForMember(potentialMemberToRemoveIP)
+		if err != nil {
+			klog.Errorf("unable to get a node for member: %v, err: %v", memberLocator, err)
+			continue
+		}
+		if node != nil {
+			klog.V(4).Infof("cannot remove member: %v because a node resource still exists (%v/%v)", memberLocator, node.Name, node.UID)
+			continue
+		}
+
+		// it looks like we have found a member that isn't backed by a machine nor a node
+		// issue a live request to the etcd cluster and remove it from the cluster
+		memberList, err := c.etcdClient.MemberList(ctx)
+		if err != nil {
+			return err
+		}
+		for _, member := range memberList {
+			memberIP, err := memberToNodeInternalIP(member)
+			if err != nil {
+				return err
+			}
+			if memberIP != potentialMemberToRemoveIP {
+				continue
+			}
+			memberLocator = fmt.Sprintf("[ url: %v, name: %v, id: %v ]", memberIP, member.Name, member.ID)
```

Comment:
I expected to see a check here to verify the member is actually unhealthy. If it's not unhealthy, you have a weird problem, but I don't think we should have the first cut here removing healthy members.

Reply:
this is a good idea, a member without a machine/node is unhealthy.